Implement bulk-move

particleflux opened the pull request:
Implement a new bulk move feature to allow moving messages between mailboxes at once.

This is able to work with an array of message ids or a MessageIterator. For compatibility with the other bulk commands, there also is the possibility to directly input a list of comma separated message ids as a string.

Fixes #305

Slamdunk commented on src/Mailbox.php:
```diff
@@ -247,4 +247,25 @@ public function getThread(): array
+            $numbers = \implode(',', $numbers);
+        }
+
+        return \imap_mail_move($this->resource->getStream(), (string) $numbers, $mailbox->getEncodedName(), \CP_UID);
```
I would like to maintain similarity with:
https://github.com/ddeboer/imap/blob/f90d03fa1a1497a00ffecb278330edfaf2c596bc/src/Message.php#L260-L262
Please remove any return and throw the Exception.

particleflux replied:
I actually was about to implement it with an exception, but decided against, as the other bulk commands (_clearFlag_ and _setFlag_) use a boolean return too.

particleflux commented:
Implemented bulk copy too. I didn't want to produce too much duplicate code, so I refactored the preparation of the message ids into a separate method. This has the nice side effect that the flag commands now work with _MessageIterators_ too.

I'm not sure if a private method is the nicest way to have that implemented. Maybe there could be some kind of _MessageList_ class for that purpose. (That would produce overhead and backward incompatible changes for the caller though `$mbox->setFlag(\Seen, new MessageList($array_of_ids)`)

Slamdunk commented on src/Mailbox.php:
```diff
@@ -247,4 +240,54 @@ public function getThread(): array
+    /**
+     * Prepare message ids for the use with bulk functions.
+     *
+     * @param array|MessageIteratorInterface|string $messageIds Message numbers
```
`getArrayCopy` isn't a method of `MessageIteratorInterface` but only of `MessageIterator`: please change all the doc hints to the latter class

particleflux replied:
right. fixed

particleflux replied:
Or not. This is now breaking the build since the static analyser complains. I could fix this by typehinting the returned variable in the test - but is that such a good idea?

```
diff --git a/tests/MailboxTest.php b/tests/MailboxTest.php
index 89cd7c5..84a58da 100644
--- a/tests/MailboxTest.php
+++ b/tests/MailboxTest.php
@@ -10,6 +10,7 @@ use Ddeboer\Imapxception\MessageDoesNotExistException;
 use Ddeboer\Imapxception\MessageMoveException;
 use Ddeboer\Imapxception\ReopenMailboxException;
 use Ddeboer\Imap\MailboxInterface;
+use Ddeboer\Imap\MessageIterator;
 
 /**
  * @covers \Ddeboer\Imapxception\AbstractException
@@ -234,6 +235,7 @@ final class MailboxTest extends AbstractTest
         $this->assertSame(0, $this->mailbox->count());
 
         // move back by iterator
+        /** @var MessageIterator $messages */
         $messages = $anotherMailbox->getMessages();
         $anotherMailbox->move($messages, $this->mailbox);
         $this->getConnection()->expunge();
```

Slamdunk replied:
> but is that such a good idea?

The bad thing is `getArrayCopy` isn't present in any SPL interface, and this complicates everything a lot.
In a future release we could add it to `MessageIteratorInterface` but at the time being your solution is enough.

Slamdunk commented on src/Mailbox.php:
```diff
@@ -247,4 +240,54 @@ public function getThread(): array
+    public function copy($numbers, MailboxInterface $mailbox)
+    {
+        if (!\imap_mail_copy($this->resource->getStream(), $this->prepareMessageIds($numbers), $mailbox->getEncodedName(), \CP_UID)) {
+            throw new MessageMoveException(\sprintf('Messages cannot be copied to "%s"', $mailbox->getName()));
```
`MessageCopyException`

particleflux replied:
Fixed too.

Slamdunk commented:
Thanks @particleflux

Slamdunk commented:
Released in [1.3.0](https://github.com/ddeboer/imap/releases/tag/1.3.0)
